File: leetcode/Longest_Common_Prefix.java
import java.util.*;

class ByLength implements Comparator<String> {
    public int compare(String a, String b) {
        if (a.length() == b.length())
            return a.compareTo(b);
        return a.length() - b.length();
    }
}

public class Solution {
    public String longestCommonPrefix(String[] strs) {
        if (strs.length == 0) return "";
        Arrays.sort(strs, new ByLength());
        String c = strs[0];
        for (int i = 1; i < strs.length; ++i) {
            if (strs[i].equals(strs[i - 1])) continue;
            int k = 0;
            while (k < c.length()) {
                if (strs[i].charAt(k) == c.charAt(k)) ++k;
                else break;
            }
            c = c.substring(0, k);
        }
        return c;
    }
}

File: leetcode/Longest_Common_Prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string> &strs) {
        if (strs.size() == 0)
            return "";
        int prefix_length = 0;
        while (true) {
            bool pass = true;
            for (int i = 0; i < strs.size(); ++i)
                if (prefix_length == strs[i].length() ||
                    strs[i][prefix_length] != strs[0][prefix_length]) {
                    pass = false;
                    break;  // Need another break to quit for-loop
                }
            if (!pass)
                break;
            ++prefix_length;
        }
        if (prefix_length == 0)
            return "";
        return strs[0].substr(0, prefix_length);
    }
};
